src/netconf_subscribed_notifications.c

```diff
@@ -572,9 +656,161 @@ csn_receiver_get_by_name(struct csn_receiver_info *recv_info, const char *name)
     return NULL;
 }
 
-int
-csn_config_sub_receivers_prepare(const struct lyd_node *node_receiver,
+/**
+ * @brief Destroy content of receiver in a subscription
+ *
+ * @param[in] receiver in the subscription in the receiver_info.
+ * @param[in] keep_ref 0: only the udp connection is destroyed, for restart, 1: everything is deleted.
+ */
+static void
+csn_receiver_destroy(struct csn_receiver *receiver, int keep_ref)
+{
+    if (!receiver) {
+        return;
+    }
+
+    if (!keep_ref) {
+        free(receiver->name);
+        receiver->name = NULL;
+
+        free(receiver->instance_ref);
+        receiver->instance_ref = NULL;
+    }
+
+    free_sender_socket(receiver->udp.sender);
+    receiver->udp.sender = NULL;
+
+    free(receiver->udp.options.address);
+    receiver->udp.options.address = NULL;
+
+    free(receiver->udp.options.port);
+    receiver->udp.options.port = NULL;
+
+    free(receiver->udp.options.interface);
+    receiver->udp.options.interface = NULL;
+
+    free(receiver->udp.options.local_address);
+    receiver->udp.options.local_address = NULL;
+}
+
+/**
+ * @brief Start a receiver to send udp notification
+ *
+ * @param[in,out] receiver the receiver to configure and start.
+ * @param[in] recv_config the configuration of this receiver.
+ * @param[in] recv_info, the receiver info structure in the subscription.
+ * @return 0 on success, -1 on error.
+ */
+static int
+csn_receiver_start(struct csn_receiver *receiver, struct csn_receiver_config *recv_config,
         struct csn_receiver_info *recv_info)
+{
+    receiver->state = CSN_RECEIVER_STATE_CONNECTING;
+    receiver->reset_time = np_gettimespec(1);
+
+    receiver->udp.options.default_mtu = 1500;
+    receiver->udp.options.address = strdup(recv_config->udp.address);
+    if (!receiver->udp.options.address) {
+        EMEM;
+        goto error;
+    }
+
+    receiver->udp.options.port = strdup(recv_config->udp.port);
+    if (!receiver->udp.options.port) {
+        EMEM;
+        goto error;
+    }
+
+    if (recv_info->local_address) {
+        receiver->udp.options.local_address = strdup(recv_info->local_address);
+        if (!receiver->udp.options.local_address) {
+            EMEM;
+            goto error;
+        }
+    }
+
+    if (recv_info->interface) {
+        receiver->udp.options.interface = strdup(recv_info->interface);
+        if (!receiver->udp.options.interface) {
+            EMEM;
+            goto error;
+        }
+    }
+
+    receiver->udp.sender = unyte_start_sender(&receiver->udp.options);
+    if (!receiver->udp.sender) {
+        ERR("Cannot create udp sender\n");
```

Comment:
Redundant `\n`, always appended. Every message should, in general, start with a capital, end with a period and ideally provide error cause if possible (such as `strerror(errno)` for standard functions, not sure if that is possible here).